File: template/src/main/java/org/maxur/mserv/events/Event.java
package org.maxur.mserv.events;

/**
 * The type Event.
 *
 * @author myunusov
 * @version 1.0
 * @since <pre>16.07.2016</pre>
 */
public abstract class Event {

    private final Object entity;

    /**
     * Instantiates a new Event.
     *
     * @param entity the entity
     */
    Event(final Object entity) {
        this.entity = entity;
    }

    /**
     * Entity t.
     *
     * @param <T> the type parameter
     * @return the t
     */
    @SuppressWarnings("unchecked")
    public <T> T entity() {
        return (T) entity;
    }
}

File: maxur-mserv-core/src/main/kotlin/org/maxur/mserv/core/command/Event.kt
package org.maxur.mserv.core.command

import org.maxur.mserv.core.Id
import java.time.Instant

/**
 * The Domain Event
 * @author Maxim Yunusov
 * @version 1.0
 * @since <pre>24.11.2017</pre>
 */
abstract class Event {

    /** This Identifier of Events. It's not mandatory property */
    var id: Id<Event> = Id.Unknown
        get() = field
        private set(value) {
            field = value
        }

    val occurredOn = Instant.now()

    /* XXX May be immutable */
    /** Identify this Event as [identifier] */
    fun identifyAs(identifier: Id<Event>): Event {
        this.id = identifier
        return this
    }
}
